pkg/controller/etcdproxy/certs.go

```diff
@@ -7,198 +7,350 @@ import (
-			secret.Namespace, secret.Name)},
-			r.Int63n(100000), currentTime)
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
-		clientCertBytes, clientKeyBytes, err := clientBundle.GetPEMBytes() // goes to apiserver
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
+	return clientCABundle.NewClientCertificate(pkix.Name{CommonName: fmt.Sprintf("client-%s-%s", clientCertSecret.Namespace, clientCertSecret.Name)},
+		r.Int63n(100000), currentTime)
+}
 
-		// Write certificate and key pair to the Secret.
-		err = c.ensureClientCert(etcdstorage, secret, clientCertBytes, clientKeyBytes, clientBundle.Certificates[0].NotAfter)
-		if err != nil {
-			errs = append(errs, err)
-		}
+// generateServerBundle generates both Serving CA bundle and Server certificate/key pair.
+func (c *EtcdProxyController) generateServerBundle(etcdstorage *etcdstoragev1alpha1.EtcdStorage) (*certs.Certificate, *certs.Certificate, error) {
+	currentTime := time.Now
+	r := rand.New(rand.NewSource(currentTime().UnixNano()))
+	serviceUrl := fmt.Sprintf("%s.%s.svc", serviceName(etcdstorage), c.config.ControllerNamespace)
+
+	// Generate the Serving CA bundle.
+	servingCA, err := certs.NewCACertificate(pkix.Name{
+		CommonName: fmt.Sprintf("%s-server-signer-%v", serviceUrl, time.Now().Unix()),
+	}, r.Int63n(100000), currentTime)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	return utilerrors.NewAggregate(errs)
+	// Generate server certificate/key pair.
+	serverCerts, err := servingCA.NewServerCertificate(pkix.Name{CommonName: serviceUrl},
+		[]string{serviceUrl}, r.Int63n(100000), currentTime)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return servingCA, serverCerts, nil
 }
 
-// ensureClientCABundle ensures the ConfigMap contains the required Client CA bundle.
-// The Client CA bundle is located in the controller namespace and is used by the etcd-proxy to verify API server identity.
-func (c *EtcdProxyController) ensureClientCABundle(etcdstorage *etcdstoragev1alpha1.EtcdStorage, caBytes []byte, expiryDate time.Time) error {
-	configMap := &v1.ConfigMap{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      etcdProxyCAConfigMapName(etcdstorage),
-			Namespace: c.config.ControllerNamespace,
-			// TODO: Remove this once validation is in place.
-			Annotations: map[string]string{
-				ProxyCertificateExpiryAnnotation: expiryDate.Format(time.RFC3339),
-			},
-			OwnerReferences: []metav1.OwnerReference{
-				*metav1.NewControllerRef(etcdstorage, etcdstoragev1alpha1.SchemeGroupVersion.WithKind("EtcdStorage")),
+// getClientCABundleConfigMap returns ConfigMap from Kube if it already exists, or returns empty ConfigMap to be populated.
+func (c *EtcdProxyController) getClientCABundleConfigMap(etcdstorage *etcdstoragev1alpha1.EtcdStorage) (*v1.ConfigMap, error) {
+	configMap, err := c.kubeclientset.CoreV1().ConfigMaps(c.config.ControllerNamespace).Get(etcdProxyCAConfigMapName(etcdstorage), metav1.GetOptions{})
+	if errors.IsNotFound(err) {
```

Comment:
just return the error and let the caller handle it

Reply:
in-lined